### PR TITLE
Added stream hashing and signing

### PR DESCRIPTION
**Pull request by aleksuss**

### CODE REVIEW

File: exonum/src/crypto.rs [hunk 1]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
     }
 }
 
+/// This structure provide possibility to calculate hash for a stream of data
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: ...structure provide**s a** possibility...

File: exonum/src/crypto.rs [hunk 2]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+///     hash_stream.update(&chunk);
+/// }
+/// let _ = hash_stream.finalize();
+/// ```
+#[derive(Debug, Default)]
+pub struct HashStream(HashState);
+
+impl HashStream {
+
+    /// Create new instance of `HashStream`
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Create**s a** new...

File: exonum/src/crypto.rs [hunk 3]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+    pub fn new() -> Self {
+        HashStream(HashState::init())
+    }
+
+    /// Processes chunk of stream using state 
+    pub fn update(&mut self, chunk: &[u8]) {
+        self.0.update(chunk);
+    }
+
+    /// Does the same as `update` but also return instance of `HashStream`
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: ...return**s**...

File: exonum/src/crypto.rs [hunk 4]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+    pub fn update_chain(mut self, chunk: &[u8]) -> Self {
+        self.0.update(chunk);
+        self
+    }
+
+    /// Complete process and return final hash of stream data
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Completes the process and returns...

File: exonum/src/crypto.rs [hunk 5]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+    pub fn new() -> Self {
+        SignStream(SignState::init())
+    }
+
+    /// Add a new `chunk` chunk to the message that will eventually be signed
+    pub fn update(&mut self, chunk: &[u8]) {
+        self.0.update(chunk);
+    }
+
+    /// Does the same as `update` but also return instance of `SignStream`
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: ...returns...

File: exonum/src/crypto.rs [hunk 6]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+
+    /// Complete process and return final hash of stream data 
+    pub fn finalize(self) -> Hash {
+        let dig = self.0.finalize();
+        Hash(dig)
+    }
+
+}
+
+/// This structure provide possibility to create signature for a stream of data
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: ...provides a possibility...

File: exonum/src/crypto.rs [hunk 7]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+/// }
+/// let file_sign = create_stream.finalize(&sk);
+/// assert!(verify_stream.verify(&file_sign, &pk));
+/// ```
+#[derive(Debug, Default)]
+pub struct SignStream(SignState);
+
+impl SignStream {
+
+    /// Create new instance of `SignStream`
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Creates a new instance of `SignStream`.

File: exonum/src/crypto.rs [hunk 8]
```diff
@@ -148,6 +148,101 @@ pub fn init() {
...
+    pub fn new() -> Self {
+        SignStream(SignState::init())
+    }
+
+    /// Add a new `chunk` chunk to the message that will eventually be signed
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Adds...

File: exonum/src/crypto.rs [hunk 9]
```diff
@@ -148,6 +148,97 @@ pub fn init() {
...
+        self.0.update(chunk);
+    }
+
+    /// Does the same as `update` but also returns instance of `SignStream`
+    pub fn update_chain(mut self, chunk: &[u8]) -> Self {
+        self.0.update(chunk);
+        self
+    }
+
+    /// Computes a signature for the previously supplied messages,
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Needless comma. 😄

File: exonum/src/crypto.rs [hunk 10]
```diff
@@ -148,6 +148,97 @@ pub fn init() {
...
+/// for chunk in data {
+///     hash_stream.update(&chunk);
+/// }
+/// let _ = hash_stream.finalize();
+/// ```
+#[derive(Debug, Default)]
+pub struct HashStream(HashState);
+
+impl HashStream {
+    /// Create a new instance of `HashStream`
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Creates...

File: exonum/src/crypto.rs [hunk 11]
```diff
@@ -148,6 +148,97 @@ pub fn init() {
     }
 }
 
+/// This structure provide a possibility to calculate hash for a stream of data
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: provides

File: exonum/src/crypto.rs [hunk 12]
```diff
@@ -148,6 +148,97 @@ pub fn init() {
...
+        self.0.update(chunk);
+    }
+
+    /// Does the same as `update` but also returns instance of `HashStream`
+    pub fn update_chain(mut self, chunk: &[u8]) -> Self {
+        self.0.update(chunk);
+        self
+    }
+
+    /// Complete process and returns final hash of stream data
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: Completes the process and returns...

File: exonum/src/crypto.rs [hunk 13]
```diff
@@ -148,6 +148,97 @@ pub fn init() {
...
+    pub fn finalize(self) -> Hash {
+        let dig = self.0.finalize();
+        Hash(dig)
+    }
+}
+
+/// This structure provide a possibility to create signature for a stream of data
```
Comment (DarkEld3r) [inline exonum/src/crypto.rs]: provides

**Comment by defuz** [comment]
@aleksuss I propose to combine methods `update` and `update_chain`. The name should be just `update`, but behavior should be "chain" :)

**Comment by defuz** [comment]
@aleksuss I also suggest renaming methods `finalize` to `hash()` and `sign()`, in accordance with `verify()`

File: exonum/src/crypto.rs [hunk 14]
```diff
@@ -148,6 +148,87 @@ pub fn init() {
...
+
+    /// Adds a new `chunk` to the message that will eventually be signed
+    pub fn update(mut self, chunk: &[u8]) -> Self {
+        self.0.update(chunk);
+        self
+    }
+
+    /// Computes a signature for the previously supplied messages
+    /// using the secret key `secret_key` and returns `Signature`
+    pub fn hash(&mut self, secret_key: &SecretKey) -> Signature {
```
Comment (defuz) [inline exonum/src/crypto.rs]: `sign()`

**Pull request merged**